tests/tb_device_mqtt_client_connect_tests.py

```diff
@@ -0,0 +1,132 @@
+        ]
+        client._subscribe_to_topic.assert_has_calls(expected_sub_calls, any_order=False)
+
+        self.assertTrue(client._TBDeviceMqttClient__request_service_configuration_required)
+
+    def test_on_connect_fail_known_code(self):
+        client = TBDeviceMqttClient("host", 1883, "username")
+        client._subscribe_to_topic = MagicMock()
+
+        known_error_code = 1
+        client._on_connect(client=None, userdata=None, flags=None, result_code=known_error_code)
+
+        self.assertFalse(client.is_connected())
+        client._subscribe_to_topic.assert_not_called()
+
+    def test_on_connect_fail_unknown_code(self):
+        client = TBDeviceMqttClient("host", 1883, "username")
+        client._subscribe_to_topic = MagicMock()
+
+        client._on_connect(client=None, userdata=None, flags=None, result_code=999)
+
+        self.assertFalse(client.is_connected())
+        client._subscribe_to_topic.assert_not_called()
+
+    def test_on_connect_fail_reasoncodes(self):
+        client = TBDeviceMqttClient("host", 1883, "username")
+        client._subscribe_to_topic = MagicMock()
+
+        mock_rc = MagicMock(spec=ReasonCodes)
+        mock_rc.getName.return_value = "SomeError"
+
+        client._on_connect(client=None, userdata=None, flags=None, result_code=mock_rc)
+
+        self.assertFalse(client.is_connected())
+        client._subscribe_to_topic.assert_not_called()
+
+    def test_on_connect_callback_with_tb_client(self):
+        client = TBDeviceMqttClient("host", 1883, "username")
+
+        def my_connect_callback(client_param, userdata, flags, rc, *args, tb_client=None):
+            self.assertIsNotNone(tb_client, "tb_client must be passed to the callback")
+            self.assertEqual(tb_client, client)
+
+        client._TBDeviceMqttClient__connect_callback = my_connect_callback
+
+        client._on_connect(client=None, userdata="test_user_data", flags="test_flags", result_code=0)
+
+    def test_on_connect_callback_without_tb_client(self):
+        client = TBDeviceMqttClient("host", 1883, "username")
+
+        def my_callback(client_param, userdata, flags, rc, *args):
+            pass
+
+        client._TBDeviceMqttClient__connect_callback = my_callback
+
+        client._on_connect(client=None, userdata="test_user_data", flags="test_flags", result_code=0)
+
+
+class TestTBDeviceMqttClient(unittest.TestCase):
+    @patch('tb_device_mqtt.paho.Client')
+    def setUp(self, mock_paho_client):
+        self.mock_mqtt_client = mock_paho_client.return_value
+        self.client = TBDeviceMqttClient(
+            host='host',
+            port=1883,
+            username='username',
+            password=None
+        )
+        self.client._TBDeviceMqttClient__service_loop = Thread(target=lambda: None)
+        self.client._TBDeviceMqttClient__updating_thread = Thread(target=lambda: None)
+
+        if not hasattr(self.client, '_client'):
```

Comment:
This check is not necessary here, because _client assigns to TBDeviceMqttClient in initialisation, so you need to overwrite it anyway to your mocked client.